README.md
# Analyzing Hi-C and HiChIP data with HiCDCPlus

Merve Sahin

05/08/2022

 A necessary task in the analysis of HiC or HiChIP count data is the detection of statistically significant and differential genomic interactions. 
  The count data are available as a table which reports, with regions typically as genomic regions binned uniformly or across restriction enzyme fragments, the number of interactions between pairs of genomic regions. The package HiCDCPlus
  provides methods to determine significant and differential chromatin interactions by use of a
  negative binomial generalized linear model, as well as implementations for TopDom to call topologically associating domains (TADs), and Juicer eigenvector to find the A/B compartments. This vignette explains the use of
  the package and demonstrates typical workflows on HiC and HiChIP data.
  HiCDCPlus package version: 0.99.16
output:
  html_document:
    keep_md: true





**Note:** if you use HiCDCPlus in published research, please cite:

> Sahin, M., Wong, W., Zhan, Y., Van Deyze, K., Koche, R., and Leslie, C. S.
> (2021)
> HiC-DC+: systematic 3D interaction calls and differential analysis 
> for Hi-C and HiChIP
> *Nature Communications*, **12(3366)**.
> [10.1038/s41467-021-23749-x](http://dx.doi.org/10.1038/s41467-021-23749-x)

# Installation

To install this package, start R and enter:


```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("HiCDCPlus")
```

If you are reinstalling the package, we recommend erasing the associated file cache for the package. The cache folder location can be obtained by running.

```r
cache <- rappdirs::user_cache_dir(appname="HiCDCPlus")
print(cache)
```

# Standard workflow

## Overview {#overview}

`HiCDCPlus` can take the outputs of the popular Hi-C pre-processing tools such as .hic (from Juicebox), .matrix, and .allValidPairs (from HiC-Pro). It can also be used with HTClist objects (from Bioconductor package HiTC). 

In the standard workflow, one first needs to generate genomic features present in the `HiCDCPlus` model (GC content, mappability, effective length) using the `construct_features` function (see [Creating genomic feature files](#bintolen)). This can be either done for uniformly or multiple restriction fragment binned data. 

`HiCDCPlus` stores counts and features in a memory-efficient way using what we name as a `gi_list` instance(see [The `gi_list` instance](#gi_list)). One next feeds the genomic features in the form of a `gi_list` instance using `generate_bintolen_gi_list` function. Then, counts can be added to this `gi_list` instance using dedicated functions for each input Hi-C file format (`add_hic_counts`, `add_hicpro_matrix_counts`,`add_hicpro_allvalidpairs.counts`).  

Before modeling, 1D features from the `gi_list` instance coming from the bintolen file must be expanded to 2D using `expand_1D_features` function. Different transformations can be applied to combine genomic features derived for each anchor. 

At the core of `HiCDCPlus` is an efficient implementation of the [HiC-DC](https://www.nature.com/articles/ncomms15454) negative binomial count model for normalization and removal of biases
(see ?HiCDCPlus). A platform-agnostic parallelizable implementation is also available in the `HiCDCPlus_parallel` function for efficient interaction 
calling across chromosomes. The `HiCDCPlus` (or `HiCDCPlus_parallel`)
function outputs the significance of each interaction (`pvalue` and FDR
adjusted p-value `qvalue`) along with following estimated from the model: 
1. `mu`: expected interaction frequency estimated from biases, 
2. `sdev`: the standard deviation of expected interaction frequencies. 

Once results are obtained, they can be output into text files using `gi_list_write` function or to a `.hic` file using the `hicdc2hic`function (where one can pass either raw counts, observed/expected normalized
counts, -log10 _P_-value, -log10 _P_-adjusted value, or 
negative binomial Z-score normalized counts: (counts-mu)/sdev to the `.hic` file

To detect differential significant interactions across conditions, `HiCDCPlus` also provides a modified implementation of
[DESeq2](https://bioconductor.org/packages/DESeq2/) using replicate Hi-C/HiChIP datasets `hicdcdiff`. This function requires a
(1) definition of the experimental setup (see ?hicdcdiff), (2) a filtered set of interactions to consider, as a text file containing columns `chr`, `startI`, and `startJ` (startI<=startJ) and (3)
count data per each condition and replicate either as `gi_list` instances or as output text files generated using the `gi_list_write` function that can be read as valid `gi_list` instances using `gi_list_read`.
The `hicdcdiff`
function performs the differential analysis and outputs genomic coordinates of
pairs of regions with corresponding logFC difference, _P_-value and BH adjusted
_P_-value (see the example in [Quickstart](#diff_int)).

We next demonstrate the standard workflow to detect significant as well as differential interactions. 

## Quickstart {#quickstart}

In this section we show a complete workflow for identifying significant
interactions and differential interactions from Hi-C data across replicate
experiments. For HiChIP, the functions used are the same, but the distance thresholds used are slightly reduced (recommended Dmax = 1.5e6).

### Finding Significant Interactions from Hi-C/HiChIP

Here, we identify significant interactions from HiC data at 50kb resolution across multiple chromosomes (in the example
below, across chromosomes 21 and 22). The following example code chunk assumes that
you have downloaded a `.hic` file from [GSE63525](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63525) and 
also downloaded [Juicebox command line tools](https://github.com/aidenlab/juicer/wiki/Download). Example
below runs with [GSE63525_HMEC_combined.hic](http://hgdownload.soe.ucsc.edu/gbdb/hg19/bbi/hic/GSE63525_HMEC_combined.hic) and stores the path to it into the variable `hicfile_path` with features generated for restriction enzyme fragments with the pattern `"GATC"` in hg19 genome.


```r
hicfile_path<-system.file("extdata", "GSE63525_HMEC_combined_example.hic", package = "HiCDCPlus")
outdir<-tempdir(check=TRUE)
#generate features
construct_features(output_path=paste0(outdir,"/hg19_50kb_GATC"),
                   gen="Hsapiens",gen_ver="hg19",
                   sig="GATC",
                   bin_type="Bins-uniform",
                   binsize=50000,
                   chrs=c("chr21","chr22"))
```

```
## [1] "Using chr21 chr22and cut patterns GATC"
## [1] "chr21"
## [1] "chr22"
```

```
## [1] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/hg19_50kb_GATC_bintolen.txt.gz"
```

If you have a multiple enzyme cocktail used to generate Hi-C data, you can specify multiple patterns including `"N"` as string to this function (e.g., sig=c("GATC","GANTC")).
If you want to analyze data binned by multiple restriction enzyme fragments, you can change bin_type to "Bins-RE-sites", and binsize to the number of fragments that you would like to merge as bin (e.g., bin_type="Bins-RE-sites" and binsize=10 means 10 restriction fragment binning).


```r
#generate gi_list instance
gi_list<-generate_bintolen_gi_list(
  bintolen_path=paste0(outdir,"/hg19_50kb_GATC_bintolen.txt.gz"))
#add .hic counts
gi_list<-add_hic_counts(gi_list,hic_path = hicfile_path)
```

```
## [1] "Chromosome chr21 intrachromosomal counts processed."
## [1] "Chromosome chr22 intrachromosomal counts processed."
```

If you have HiC-Pro outputs instead, you can use either `add_hicpro_matrix_counts` or `add_hicpro_allvalidpairs.counts` depending on the file format. `add_hicpro_matrix_counts` function requires .bed output from HiC-Pro matrix generation step, together with count data in .matrix format.


```r
#expand features for modeling
gi_list<-expand_1D_features(gi_list)
#run HiC-DC+ on 2 cores
set.seed(1010) #HiC-DC downsamples rows for modeling
gi_list<-HiCDCPlus_parallel(gi_list,ncore=2)
head(gi_list)
```

```
## $chr21
## GInteractions object with 18026 interactions and 9 metadata columns:
##           seqnames1           ranges1     seqnames2           ranges2 |
##               <Rle>         <IRanges>         <Rle>         <IRanges> |
##       [1]     chr21   9450000-9500000 ---     chr21  9950000-10000000 |
##       [2]     chr21   9450000-9500000 ---     chr21 10100000-10150000 |
##       [3]     chr21   9450000-9500000 ---     chr21 10150000-10200000 |
##       [4]     chr21   9450000-9500000 ---     chr21 11000000-11050000 |
##       [5]     chr21   9450000-9500000 ---     chr21 11100000-11150000 |
##       ...       ...               ... ...       ...               ... .
##   [18022]     chr21 47900000-47950000 ---     chr21 47900000-47950000 |
##   [18023]     chr21 47900000-47950000 ---     chr21 47950000-48000000 |
##   [18024]     chr21 47900000-47950000 ---     chr21 48000000-48050000 |
##   [18025]     chr21 47900000-47950000 ---     chr21 48050000-48100000 |
##   [18026]     chr21 48000000-48050000 ---     chr21 48000000-48050000 |
##                   D    counts        gc       map       len        mu      sdev
##           <integer> <numeric> <numeric> <numeric> <numeric> <numeric> <numeric>
##       [1]    500000         0 -0.415451         0 0.2861372  36.62682  20.85646
##       [2]    650000         0 -0.767405         0 0.2689031  29.08634  16.74247
##       [3]    700000         0 -0.461107         0 0.2347835  25.74290  14.91746
##       [4]   1550000         0 -0.795873         0 0.0742543   7.95680   5.17270
##       [5]   1650000         0  0.299047         0 0.2230742   8.34009   5.38471
##       ...       ...       ...       ...       ...       ...       ...       ...
##   [18022]         0      2200  0.633416         0 0.4511124  1789.121  975.8642
##   [18023]     50000       545  0.752876         0 0.1295933   537.613  293.8783
##   [18024]    100000       260  0.725338         0 0.3141704   276.445  151.5580
##   [18025]    150000       271  0.570906         0 0.0480836   125.598   69.3535
##   [18026]         0      2709  0.817260         0 0.1772284  1484.744  809.9996
##              pvalue    qvalue
##           <numeric> <numeric>
##       [1]         1         1
##       [2]         1         1
##       [3]         1         1
##       [4]         1         1
##       [5]         1         1
##       ...       ...       ...
##   [18022] 0.2837514  0.857618
##   [18023] 0.4184339  0.922653
##   [18024] 0.4722949  0.942038
##   [18025] 0.0386625  0.387830
##   [18026] 0.0811332  0.562719
##   -------
##   regions: 963 ranges and 3 metadata columns
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $chr22
## GInteractions object with 18241 interactions and 9 metadata columns:
##           seqnames1           ranges1     seqnames2           ranges2 |
##               <Rle>         <IRanges>         <Rle>         <IRanges> |
##       [1]     chr22 16050000-16100000 ---     chr22 17650000-17700000 |
##       [2]     chr22 16050000-16100000 ---     chr22 17700000-17750000 |
##       [3]     chr22 16050000-16100000 ---     chr22 17800000-17850000 |
##       [4]     chr22 16050000-16100000 ---     chr22 17950000-18000000 |
##       [5]     chr22 16050000-16100000 ---     chr22 18000000-18050000 |
##       ...       ...               ... ...       ...               ... .
##   [18237]     chr22 51000000-51050000 ---     chr22 51100000-51150000 |
##   [18238]     chr22 51000000-51050000 ---     chr22 51150000-51200000 |
##   [18239]     chr22 51050000-51100000 ---     chr22 51050000-51100000 |
##   [18240]     chr22 51050000-51100000 ---     chr22 51150000-51200000 |
##   [18241]     chr22 51150000-51200000 ---     chr22 51150000-51200000 |
##                   D    counts         gc       map       len        mu
##           <integer> <numeric>  <numeric> <numeric> <numeric> <numeric>
##       [1]   1600000         1  0.0731079         0 0.1565584   6.52084
##       [2]   1650000         3  0.0445727         0 0.0809363   5.78804
##       [3]   1750000         3 -0.2745115         0 0.0120789   4.57491
##       [4]   1900000         0 -0.6320482         0 0.0357296   3.31760
##       [5]   1950000         1 -0.2308528         0 0.1088195   3.23818
##       ...       ...       ...        ...       ...       ...       ...
##   [18237]    100000       477   1.350818         0 0.1560164   414.965
##   [18238]    150000       269   1.113756         0 0.3827365   246.312
##   [18239]         0      3903   0.440179         0 0.1683074  2610.342
##   [18240]    100000       350   0.540858         0 0.0930645   361.689
##   [18241]         0      2665   0.641536         0 0.0178216  2415.709
##                sdev    pvalue    qvalue
##           <numeric> <numeric> <numeric>
##       [1]   4.31807  0.974889         1
##       [2]   3.91674  0.795263         1
##       [3]   3.24700  0.704750         1
##       [4]   2.54088  1.000000         1
##       [5]   2.49563  0.899118         1
##       ...       ...       ...       ...
##   [18237]   222.523  0.329603  1.000000
##   [18238]   132.462  0.366718  1.000000
##   [18239]  1394.840  0.163675  0.950785
##   [18240]   194.074  0.453822  1.000000
##   [18241]  1290.907  0.357891  1.000000
##   -------
##   regions: 1027 ranges and 3 metadata columns
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```

```r
#write normalized counts (observed/expected) to a .hic file
hicdc2hic(gi_list,hicfile=paste0(outdir,'/GSE63525_HMEC_combined_result.hic'),
          mode='normcounts',gen_ver='hg19')
```

```
## [1] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/GSE63525_HMEC_combined_result.hic"
```

```r
#write results to a text file
gi_list_write(gi_list,fname=paste0(outdir,'/GSE63525_HMEC_combined_result.txt.gz'))
```

```
## [1] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/GSE63525_HMEC_combined_result.txt.gz"
```

`HiCDCPlus` results can be converted into .hic using `hicdc2hic` function. Values that should be supplied as "mode" into the `hicdc2hic` function for the corresponding score stored in the .hic file are: 'pvalue' for -log10 significance p-value, 'qvalue' for -log10 FDR corrected p-value, 'normcounts' for raw counts/expected counts, 'zvalue' for standardized counts (raw counts-expected counts)/modeled standard deviation of expected counts and 'raw' to pass-through raw counts. 

.hic files can be further converted into .cool format using hic2cool software and be visualized using HiCExplorer. 

## Finding Differential Interactions {#diff_int}

Suppose we're interested in finding differential interactions on `chr21` 
and `chr22` at 50kb between
NSD2 and NTKO/TKO cells given the following `.hic` files available in
[GSE131651](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE131651):
`GSE131651_NSD2_LOW_arima.hic`, `GSE131651_NSD2_HIGH_arima.hic`,
`GSE131651_TKOCTCF_new.hic`, `GSE131651_NTKOCTCF_new.hic`. We first find 
significant interactions in each, and save results to a file:


```r
#generate features
construct_features(output_path=paste0(outdir,"/hg38_50kb_GATC"),
                   gen="Hsapiens",gen_ver="hg38",
                   sig="GATC",bin_type="Bins-uniform",
                   binsize=50000,
                   chrs=c("chr21","chr22"))
```

```
## [1] "Using chr21 chr22and cut patterns GATC"
## [1] "chr21"
## [1] "chr22"
```

```
## [1] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/hg38_50kb_GATC_bintolen.txt.gz"
```

```r
#add .hic counts
hicfile_paths<-c(
system.file("extdata", "GSE131651_NSD2_LOW_arima_example.hic", package = "HiCDCPlus"),
system.file("extdata", "GSE131651_NSD2_HIGH_arima_example.hic", package = "HiCDCPlus"),
system.file("extdata", "GSE131651_TKOCTCF_new_example.hic", package = "HiCDCPlus"),
system.file("extdata", "GSE131651_NTKOCTCF_new_example.hic", package = "HiCDCPlus"))
indexfile<-data.frame()
for(hicfile_path in hicfile_paths){
output_path<-paste0(outdir,'/',
                    gsub("^(.*[\\/])", "",gsub('.hic','.txt.gz',hicfile_path)))
#generate gi_list instance
gi_list<-generate_bintolen_gi_list(
  bintolen_path=paste0(outdir,"/hg38_50kb_GATC_bintolen.txt.gz"),
  gen="Hsapiens",gen_ver="hg38")
gi_list<-add_hic_counts(gi_list,hic_path = hicfile_path)
#expand features for modeling
gi_list<-expand_1D_features(gi_list)
#run HiC-DC+ on 2 cores
set.seed(1010) #HiC-DC downsamples rows for modeling
gi_list<-HiCDCPlus(gi_list,ssize=0.1)
for (i in seq(length(gi_list))){
indexfile<-unique(rbind(indexfile,
  as.data.frame(gi_list[[i]][gi_list[[i]]$qvalue<=0.05])[c('seqnames1',
                                                           'start1','start2')]))
}
#write results to a text file
gi_list_write(gi_list,fname=output_path)
}
```

```
## Warning in add_2D_features(gi_list[[chrom]], count_matrix): Bins and counts mismatch. This will slow down the
##             performance of counts integration.Check if genome and/or bin size of
##             counts data is aligned with the GenomicInteractions object.
```

```
## [1] "Chromosome chr21 intrachromosomal counts processed."
```

```
## Warning in add_2D_features(gi_list[[chrom]], count_matrix): Bins and counts mismatch. This will slow down the
##             performance of counts integration.Check if genome and/or bin size of
##             counts data is aligned with the GenomicInteractions object.
```

```
## [1] "Chromosome chr22 intrachromosomal counts processed."
## [1] "Chromosome chr21 complete."
## [1] "Chromosome chr22 complete."
## [1] "Chromosome chr21 intrachromosomal counts processed."
## [1] "Chromosome chr22 intrachromosomal counts processed."
## [1] "Chromosome chr21 complete."
## [1] "Chromosome chr22 complete."
## [1] "Chromosome chr21 intrachromosomal counts processed."
## [1] "Chromosome chr22 intrachromosomal counts processed."
## [1] "Chromosome chr21 complete."
## [1] "Chromosome chr22 complete."
## [1] "Chromosome chr21 intrachromosomal counts processed."
## [1] "Chromosome chr22 intrachromosomal counts processed."
## [1] "Chromosome chr21 complete."
## [1] "Chromosome chr22 complete."
```

```r
#save index file---union of significants at 50kb
colnames(indexfile)<-c('chr','startI','startJ')
data.table::fwrite(indexfile,
            paste0(outdir,'/GSE131651_analysis_indices.txt.gz'),
            sep='\t',row.names=FALSE,quote=FALSE)
```

We next get the union set of significant interactions and save it as the index file, and then run `hicdcdiff`.


```r
#Differential analysis using modified DESeq2 (see ?hicdcdiff)
hicdcdiff(input_paths=list(NSD2=c(paste0(outdir,'/GSE131651_NSD2_LOW_arima_example.txt.gz'),
                 paste0(outdir,'/GSE131651_NSD2_HIGH_arima_example.txt.gz')),
TKO=c(paste0(outdir,'/GSE131651_TKOCTCF_new_example.txt.gz'),
paste0(outdir,'/GSE131651_NTKOCTCF_new_example.txt.gz'))),
filter_file=paste0(outdir,'/GSE131651_analysis_indices.txt.gz'),
output_path=paste0(outdir,'/diff_analysis_example/'),
fitType = 'mean',
binsize=50000,
diagnostics=TRUE)
```

```
## $deseq2paths
## NULL
## 
## $outputpaths
## [1] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/diff_resTKOoverNSD2_chr21.txt.gz"
## [2] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/diff_resTKOoverNSD2_chr22.txt.gz"
## 
## $plotpaths
##  [1] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/sizefactors_chr21.pdf"        
##  [2] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/geomean_sizefactors_chr21.pdf"
##  [3] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/plotMA_TKOoverNSD2_chr21.pdf" 
##  [4] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/diff_chr21_PCA.pdf"           
##  [5] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/dispersionplot.pdf"           
##  [6] "/var/folders/31/gz4s1kx132l4_hvgmfc1jl2m0000gn/T//RtmpUrJ86Z/diff_analysis_example/sizefactors_chr22.pdf"        

inst/NEWS
Changes in version 0.99.17 (2022-06-06)
+ Fixed bug that prevented juicer to setup
Changes in version 0.99.16 (2022-05-08)
+ Updated references to juicer
Changes in version 0.99.14 (2021-04-13)
+ Fixed bug that prevents passing negative values to .hic files
Changes in version 0.99.13 (2021-03-23)
+ Added RE-agnostic features into construct_features
Changes in version 0.99.12 (2021-02-19)
+ Submitted to Bioconductor

R/hicdcdiff.R
#' hicdcdiff
#'
#'This function calculates differential interactions for a
#'set of chromosomes across conditions and replicates. You need to install
#'\code{DESeq2} from Bioconductor to use this function.
#'@importFrom dplyr %>%
#'@importFrom rlang .data
#'@param input_paths a list with names as condition names and values
#'as paths to \code{gi_list} RDS objects (see \code{?gi_list_validate}
#'for a detailed explanation of valid \code{gi_list} instances) 
#'saved with \code{saveRDS} or
#'paths to .hic files for each replicate. e.g.,\code{list(
#'CTCF=c('~/Downloads/GM_CTCF_rep1_MAPQ30_10kb.rds',
#''~/Downloads/GM_CTCF_rep2_MAPQ30_10kb.rds'),
#'SMC=c('~/Downloads/GM_SMC_rep1_MAPQ30_10kb.rds',
#''~/Downloads/GM_SMC_rep2_MAPQ30_10kb.rds'))}
#'@param filter_file path to the text file containing columns 
#'chr', startI, and
#'startJ denoting the name of the chromosomes and starting coordinates
#'of 2D interaction bins to be compared across conditions, respectively.
#'@param output_path the path to the folder and name prefix you want to
#'place DESeq-processed matrices (in a .txt file), plots
#'(if \code{diagnostics=TRUE}) and DESeq2 objects (if \code{DESeq.save=TRUE}).
#'Files will be generated for each chromosome.
#'@param bin_type 'Bins-uniform' if uniformly binned by binsize in
#'bp, or 'Bins-RE-sites' if binned by number of
#'restriction enzyme fragment cutsites!
#'@param binsize binsize in bp if bin_type='Bins-uniform' (or number of
#'RE fragments if bin_type='Bins-RE-sites'), e.g., default 5000
#'@param granularity Desired distance granularity to base dispersion parameters
#'on in bp. For uniformly binned analysis 
#'(i.e., \code{bin_type=='Bins-uniform'}),
#'this defaults to the bin size. Otherwise, it is 5000.
#'@param chrs select a subset of chromosomes' e.g.,
#'c('chr21','chr22'). Defaults to all chromosomes (except Y and M)
#'in the filter_file.
#'@param Dmin minimum distance (included) to check for significant interactions,
#'defaults to 0. Put Dmin=1 to ignore D=0 bins in calculating
#'normalization factors.
#'@param Dmax maximum distance (included) to check for significant interactions,
#' defaults to 2e6 or maximum in the data; whichever is minimum.
#'@param diagnostics if TRUE, generates diagnostic plots of the normalization
#'factors, geometric means of such factors by distance bin, as well as MA Plots
#'(see DESeq documentation for details about MA plots). Defaults to FALSE.
#'@param DESeq.save if TRUE, saves the DESeq objects for each chromosome
#'as an .rds file in the \code{output_path}. Defaults to FALSE.
#'@param fitType follows fitType in \code{DESeq2::estimateDispersions}.
#'Allowable options are 'parametric' (parametric regression),'local' 
#'(local regression), and 'mean' (constant across interaction bins). 
#'Default is 'local'.
#'@return paths of a list of three entities.
#'\code{outputpaths} will have differential bins among those in filter_file.
#'\code{deseq2paths} will have the DESeq2 object stored as an .rds file.
#'Available if \code{DESeq.save=TRUE}
#'\code{plotpaths} will have diagnostic plots (e.g., MA, dispersion, PCA)
#'if \code{diagnostics=TRUE}.
#'@examples 
#'outputdir<-paste0(tempdir(check=TRUE),'/')
#'hicdcdiff(input_paths=list(NSD2=c(
#'system.file("extdata", "GSE131651_NSD2_LOW_arima_example.hic",
#' package = "HiCDCPlus"),
#' system.file("extdata", "GSE131651_NSD2_HIGH_arima_example.hic",
#' package = "HiCDCPlus")),
#' TKO=c(system.file("extdata", "GSE131651_TKOCTCF_new_example.hic",
#' package = "HiCDCPlus"),
#' system.file("extdata", "GSE131651_NTKOCTCF_new_example.hic",
#' package = "HiCDCPlus"))),
#' filter_file=system.file("extdata", "GSE131651_analysis_indices.txt.gz",
#' package = "HiCDCPlus"),
#'          chrs='chr22',
#'          output_path=outputdir,
#'          fitType = 'mean',
#'          binsize=50000,
#'          diagnostics=FALSE)
#'@export

hicdcdiff <- function(input_paths, filter_file, output_path, bin_type = "Bins-uniform", binsize = 5000, granularity = 5000, 
    chrs = NULL, Dmin = 0, Dmax = 2e+06, diagnostics = FALSE, DESeq.save = FALSE, fitType = "local") {
    options(scipen = 100, digits = 4, warn = -1)
    conds <- names(input_paths)
    granularity <- ifelse(bin_type == "Bins-uniform" & granularity == 5000, binsize, granularity)
    dband <- seq(0, Dmax, granularity)
    dband[1] <- Dmin
    dband <- unique(sort(dband[dband >= Dmin]))
    deseq2paths <- NULL
    outputpaths <- NULL
    plotpaths <- NULL
    sigs <- data.table::fread(filter_file)
    #set memory limit to max if i386
    if (.Platform$OS.type=='windows'&Sys.getenv("R_ARCH")=="/i386") {
        gc(reset=TRUE,full=TRUE)
        utils::memory.limit(size=4095)
    }
    # set default chrs if need be
    if (is.null(chrs)) {
        # get list of chromosomes from sigs
        if (!"chr" %in% colnames(sigs)) 
            stop("No column named 'chr' in filter_file.")
        chrs <- sort(unique(sigs$chr))
    }

    for (chr in chrs) {
        plotpaths_add <- NULL
        outputpaths_add <- NULL
        deseq2paths_add <- NULL
        retlist <- .readDiffInputFiles(conds, input_paths, sigs, chr, Dmin, Dmax, dband, bin_type, binsize)
        if (diagnostics) {
            plotpaths_add <- .plotNormalizationFactors(retlist[["normfac.final"]], binsize, chr, dband, conds, output_path)
        }
        countcols <- c()
        for (cond in conds) {
            countcols <- c(countcols, paste0(cond, ".", seq(length(input_paths[[cond]]))))
        }
        count.mat <- retlist[["normfac"]][countcols]
        rownames(count.mat) <- paste0(retlist[["normfac"]]$startI, ":", retlist[["normfac"]]$startJ)
        normcols <- paste0(countcols, ".norm")
        normfac.mat <- retlist[["normfac"]][normcols]
        rownames(normfac.mat) <- rownames(count.mat)
        sampleinfo <- data.frame(condition = rep(conds, times = unlist(lapply(input_paths, length))))
        rownames(sampleinfo) <- colnames(count.mat)
        sampleinfo$samples <- rownames(sampleinfo)
        count.mat <- as.matrix(count.mat)
        normfac.mat <- as.matrix(normfac.mat)
        ## run deseq2
        dds <- DESeq2::DESeqDataSetFromMatrix(countData = count.mat, colData = sampleinfo, design = ~condition)
        dds$condition <- stats::relevel(dds$condition, ref = conds[1])
        DESeq2::normalizationFactors(dds) <- normfac.mat
        keep <- rownames(DESeq2::counts(dds)) %in% retlist[["sigbins"]]
        dds <- dds[keep, ]
        dds <- DESeq2::estimateDispersionsGeneEst(dds)
        DESeq2::dispersions(dds) <- tryCatch({
            dds <- DESeq2::estimateDispersionsFit(dds, fitType = fitType)
            GenomicRanges::mcols(dds)$dispFit},
            error=function(e) {
                msg<-paste0("fitType failed to fit for ",chr,". Overriding with gene estimates")
                message(msg)
                return(mcols(dds)$dispGeneEst)
            })
        dds <- DESeq2::nbinomWaldTest(dds)
        for (i in seq(1, length(conds) - 1, 1)) {
            for (j in seq(i + 1, length(conds), 1)) {
                # create contrast variables
                var <- paste0("res", conds[j], "over", conds[i])
                assign(var, DESeq2::results(dds, contrast = c("condition", conds[j], conds[i])))
                if (diagnostics) {
                # MA plot
                maplot_path <- path.expand(paste0(output_path, "plotMA_", conds[j], "over", conds[i], "_", chr, ".pdf"))
                maplot_pathdir<-gsub("/[^/]+$", "",maplot_path)
                if (maplot_pathdir==maplot_path){
                    maplot_pathdir<-gsub("\\[^\\]+$", "",maplot_path)
                }
                if (maplot_pathdir==maplot_path){
                    maplot_pathdir<-gsub("\\\\[^\\\\]+$", "",maplot_path)
                }
                if (!maplot_pathdir==maplot_path&!dir.exists(maplot_pathdir)){
                    dir.create(maplot_pathdir, showWarnings = FALSE, recursive = TRUE, mode = "0777")
                }
                grDevices::pdf(maplot_path)
                DESeq2::plotMA(get(var))
                grDevices::dev.off()
                plotpaths_add <- c(plotpaths_add, maplot_path)
                }
                # write to file
                assign(var, as.data.frame(get(var)))
                assign(var, get(var) %>% tibble::rownames_to_column("bin") %>% dplyr::mutate(chr = chr) %>% tidyr::separate(col = .data$bin, 
                into = c("startI", "startJ"), sep = ":", remove = TRUE, convert = TRUE))
                assign(var, get(var)[c("chr", setdiff(names(get(var)), "chr"))])
                filepath <- paste0(output_path, "diff_", var, "_", chr, ".txt.gz")
                filepath<-path.expand(filepath)
                filepathdir<-gsub("/[^/]+$", "",filepath)
                if (filepathdir==filepath){
                    filepathdir<-gsub("\\[^\\]+$", "",filepath)
                }
                if (filepathdir==filepath){
                    filepathdir<-gsub("\\\\[^\\\\]+$", "",filepath)
                }
                if (!filepathdir==filepath&!dir.exists(filepathdir)){
                    dir.create(filepathdir, showWarnings = FALSE, recursive = TRUE, mode = "0777")
                }
                data.table::fwrite(get(var), filepath, quote = FALSE, sep = "\t", row.names = FALSE)
                outputpaths_add <- c(outputpaths_add, filepath)
            }
        }
        
        if (diagnostics) {
            # PCA plot
            rld <- DESeq2::rlog(dds, fitType = fitType)
            pcaplot_path <- path.expand(paste0(output_path, "diff_", chr, "_PCA.pdf"))
            pcaplot_pathdir<-gsub("/[^/]+$", "",pcaplot_path)
            if (pcaplot_pathdir==pcaplot_path){
                pcaplot_pathdir<-gsub("\\[^\\]+$", "",pcaplot_path)
            }
            if (pcaplot_pathdir==pcaplot_path){
                pcaplot_pathdir<-gsub("\\\\[^\\\\]+$", "",pcaplot_path)
            }
            if (!pcaplot_pathdir==pcaplot_path&!dir.exists(pcaplot_pathdir)){
                dir.create(pcaplot_pathdir, showWarnings = FALSE, recursive = TRUE, mode = "0777")
            }
            grDevices::pdf(pcaplot_path)
            p <- DESeq2::plotPCA(rld, intgroup = "samples")
            print(p)
            grDevices::dev.off()
            plotpaths_add <- c(plotpaths_add, pcaplot_path)
            # Dispersion plot
            dispplot_path <- path.expand(paste0(output_path, "dispersionplot.pdf"))
            dispplot_pathdir<-gsub("/[^/]+$", "",dispplot_path)
            if (dispplot_pathdir==dispplot_path){
                dispplot_pathdir<-gsub("\\[^\\]+$", "",dispplot_path)
            }
            if (dispplot_pathdir==dispplot_path){
                dispplot_pathdir<-gsub("\\\\[^\\\\]+$", "",dispplot_path)
            }
            if (!dispplot_pathdir==dispplot_path&!dir.exists(dispplot_pathdir)){
                dir.create(dispplot_pathdir, showWarnings = FALSE, recursive = TRUE, mode = "0777")
            }
            grDevices::pdf(dispplot_path)
            DESeq2::plotDispEsts(dds)
            grDevices::dev.off()
            plotpaths_add <- c(plotpaths_add, dispplot_path)
        }
        
        # save DESeq file
        if (DESeq.save) {
            deseq2path <- paste0(output_path, chr, "_DESeq2_obj.rds")
            deseq2output <- path.expand(deseq2path)
            deseq2outputdir<-gsub("/[^/]+$", "",deseq2output)
            if (deseq2outputdir==deseq2output){
            deseq2outputdir<-gsub("\\[^\\]+$", "",deseq2output)
            }
            if (deseq2outputdir==deseq2output){
            deseq2outputdir<-gsub("\\\\[^\\\\]+$", "",deseq2output)
            }
            if (!deseq2outputdir==deseq2output&!dir.exists(deseq2outputdir)){
            dir.create(deseq2outputdir, showWarnings = FALSE, recursive = TRUE, mode = "0777")
            }
            saveRDS(dds, deseq2path)
            deseq2paths_add <- c(deseq2paths_add, deseq2path)
        }
        deseq2paths <- c(deseq2paths, deseq2paths_add)
        outputpaths <- c(outputpaths, outputpaths_add)
        plotpaths <- c(plotpaths, plotpaths_add)
    }
    return(list(deseq2paths = deseq2paths, outputpaths = outputpaths, plotpaths = plotpaths))
}
